genie-web/src/main/java/com/netflix/genie/web/services/JobKillService.java

```diff
@@ -52,4 +53,15 @@ void killJob(@NotBlank(message = "No id entered. Unable to kill job.") final Str
      */
     @EventListener
     void onKillJobEvent(@NotNull final KillJobEvent event) throws GenieException;
+
+    /**
+     * Register a callback for killing the job.
+     *
+     * @param id     id of the job
+     * @param consumer consumer representing the callback to be executed
+     */
+
+    void registerCallbackForKill(
```

Comment:
I don't necessarily agree with making this a hard method of this class. This interface is to an API for simply killing jobs. Exposing a mechanism for registering callbacks exposes an internal detail. It may well be a valuable purpose but I think we could break it out into a separate interface. Then within your gRPC service you can instead inject a `List<KillCallbackListencers>` that way we can have multiple things with the ability to register. That explanation might not be clear. My brain isn't working at 100% at the moment haha.

Effectively there may be some value of providing a way for multiple services to do this or even better we limit the methods exposed based on which interface we inject into the consumers. e.g. Within your grpc service endpoint you only inject the interface that exposes this method. It shouldn't know about any other methods or that the service kills things. It only cares "give me something I can get notifications from".

Additionally we may have over complicated it if we want it totally decoupled we could use events which we already pub/sub and not reinvent this wheel. We can talk about it offline next week when @mprimi is back

This comment is a mess of thoughts so I'm going to cut it off now

Reply:
I have created a new interface - `JobKillCallbackService` and moved the methods to register and deregister callbacks there. `LocalJobKillServiceImpl` now implements this interface in addition `JobKillService`. I inject `JobKillCallbackService` in `GrpcJobTerminationServiceImpl`